### PR TITLE
One-to-one test spec support

### PR DESCRIPTION
**Pull request by dnkoutso**
As of 1.5.x CocoaPods merges all test specs into a single test bundle target. This can cause problems when an author wants have different organization of their test sources but also include different dependencies and different configurations (e.g. one test spec requires an app host while the other doesn't).

closes https://github.com/CocoaPods/CocoaPods/issues/7573

This is an enhancement for 1.6.0 where now each test spec will get its own test bundle.

### CODE REVIEW

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* [WIP] One-to-one test spec support  
  [dnkoutso](https://github.com/dnkoutso)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 1]
```diff
@@ -445,14 +449,14 @@ def create_test_target_copy_resources_script(test_type)
...
           #
-          def create_test_target_embed_frameworks_script(test_type)
-            path = target.embed_frameworks_script_path_for_test_type(test_type)
-            pod_targets = target.all_dependent_targets
+          def create_test_target_embed_frameworks_script(test_spec)
+            path = target.embed_frameworks_script_path_for_test_spec(test_spec)
+            pod_targets = target.dependent_targets_for_test_spec(test_spec)
             framework_paths_by_config = target.user_build_configurations.keys.each_with_object({}) do |config, paths_by_config|
               paths_by_config[config] = pod_targets.flat_map do |pod_target|
                 spec_paths_to_include = pod_target == target ? pod_target.specs.map(&:name) : pod_target.non_test_specs.map(&:name)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: will need the same change as above here too.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_integrator.rb [hunk 2]
```diff
@@ -71,10 +72,9 @@ def add_copy_resources_script_phase(native_target)
...
           # @return [void]
           #
-          def add_embed_frameworks_script_phase(native_target)
-            test_type = target.test_type_for_product_type(native_target.symbol_type)
-            script_path = "${PODS_ROOT}/#{target.embed_frameworks_script_path_for_test_type(test_type).relative_path_from(target.sandbox.root)}"
-            framework_paths = target.all_dependent_targets.flat_map do |dependent_target|
+          def add_embed_frameworks_script_phase(native_target, test_spec)
+            script_path = "${PODS_ROOT}/#{target.embed_frameworks_script_path_for_test_spec(test_spec).relative_path_from(target.sandbox.root)}"
+            framework_paths = target.dependent_targets_for_test_spec(test_spec).flat_map do |dependent_target|
               spec_paths_to_include = dependent_target == target ? dependent_target.specs.map(&:name) : dependent_target.non_test_specs.map(&:name)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_integrator.rb]: same here as above.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 3]
```diff
@@ -424,17 +426,19 @@ def create_test_xcconfig_files(test_native_targets, test_resource_bundle_targets
...
-          def create_test_target_copy_resources_script(test_type)
-            path = target.copy_resources_script_path_for_test_type(test_type)
-            pod_targets = target.all_dependent_targets
+          def create_test_target_copy_resources_script(test_spec)
+            path = target.copy_resources_script_path_for_test_spec(test_spec)
+            pod_targets = target.dependent_targets_for_test_spec(test_spec)
             resource_paths_by_config = target.user_build_configurations.keys.each_with_object({}) do |config, resources_by_config|
               resources_by_config[config] = pod_targets.flat_map do |pod_target|
-                spec_paths_to_include = pod_target == target ? pod_target.specs.map(&:name) : pod_target.non_test_specs.map(&:name)
+                spec_paths_to_include = pod_target == target ? pod_target.specs.select do |spec|
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: this logic is a bit hard to follow, should we extract a method out to simplify it?

Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: it's all been used in many places.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_integrator.rb [hunk 4]
```diff
@@ -25,8 +25,8 @@ def initialize(target_installation_result)
           def integrate!
             UI.section(integration_message) do
               target_installation_result.test_specs_by_native_target.each do |test_native_target, test_specs|
-                add_embed_frameworks_script_phase(test_native_target)
-                add_copy_resources_script_phase(test_native_target)
+                add_embed_frameworks_script_phase(test_native_target, test_specs.first)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_integrator.rb]: I think this might be wrong when I first did it. I dont think we should be passing only the `test_specs.first` here.

File: lib/cocoapods/target/pod_target.rb [hunk 5]
```diff
@@ -356,55 +347,50 @@ def resources_bundle_target_label(bundle_name)
...
+    # @param  [Specification] test_spec
+    #         The test spec to use for producing the test label.
     #
     # @return [String] The derived name of the test target.
     #
-    def test_target_label(test_type)
-      "#{label}-#{test_type.capitalize}-Tests"
+    def test_target_label(test_spec)
+      # TODO do not use base_name here
+      "#{label}-#{test_spec.test_type.capitalize}-#{test_spec.base_name}"
```
Comment (dnkoutso) [inline lib/cocoapods/target/pod_target.rb]: @segiddins you and I had talked about this, what is the best thing to use to make sure the name of a test target is uniq?

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb (hunk 3 above)
Comment (jkap) [reply lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: probably a good call

File: lib/cocoapods/target/pod_target.rb (hunk 5 above)
Comment (dnkoutso) [reply lib/cocoapods/target/pod_target.rb]: also is this method used outside of `pod_target_installer.rb`? If no then we can move it there?

Comment (dnkoutso) [reply lib/cocoapods/target/pod_target.rb]: nevermind on last question seems like it is so we keep it here.

**Comment by dnkoutso** [comment]
If not in this PR, we should fast follow with a integration test that includes multiple test specs and various different combos to ensure it builds and tests correctly.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 6]
```diff
@@ -414,13 +416,13 @@ def create_test_xcconfig_files(test_native_targets, test_resource_bundle_targets
             target.test_specs.each do |test_spec|
               spec_consumer = test_spec.consumer(target.platform)
               test_type = spec_consumer.test_type
-              path = target.xcconfig_path(test_type.to_s)
+              path = target.xcconfig_path("#{test_type.capitalize}-#{test_spec.base_name}")
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: there could in theory be multiple test specs with the same base name?

File: lib/cocoapods/target/pod_target.rb [hunk 7]
```diff
@@ -356,55 +346,50 @@ def resources_bundle_target_label(bundle_name)
...
-    #         The test type to use for producing the test label.
+    # @param  [Specification] test_spec
+    #         The test spec to use for producing the test label.
     #
     # @return [String] The derived name of the test target.
     #
-    def test_target_label(test_type)
-      "#{label}-#{test_type.capitalize}-Tests"
+    def test_target_label(test_spec)
+      # TODO: do not use base_name here
```
Comment (segiddins) [inline lib/cocoapods/target/pod_target.rb]: TODO

Comment (dnkoutso) [reply lib/cocoapods/target/pod_target.rb]: What is the best thing to use here? is `base_name` uniq enough?

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb (hunk 6 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: true we should figure out something uniq

File: CHANGELOG.md [hunk 8]
```diff
@@ -8,6 +8,10 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 
 ##### Enhancements
 
+* One-to-one test spec support
```
Comment (segiddins) [inline CHANGELOG.md]: `Every test spec will result in its own xctest bundle` ?

File: CHANGELOG.md [hunk 9]
```diff
@@ -8,6 +8,10 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 
 ##### Enhancements
 
+* One-to-one test spec support  
+  [Dimitris Koutsogiorgas](https://github.com/dnkoutso) and [Jenn Kaplan](https://github.com/jkap)
```
Comment (segiddins) [inline CHANGELOG.md]: each other goes on their own line

**Comment by jkap** [comment]
@dnkoutso finally went green! 🚢

**Comment by jverkoey** [comment]
Hey gang - this is great work, thank you for splitting these targets up.

One downside of this change is that each test target needs to be added to a main scheme used for day-to-day development so that all of the tests can be ran with a single invocation. Prior to this change it was possible to add the single generated unit test target to a scheme.

Is there work planned out to make it easier to roll these smaller targets into a scheme, or into a single meta testing target that runs all of the individual tests?

**Comment by segiddins** [comment]
The library should still have all of its test specs set us as testables in its scheme
